algo/src/main/java/org/neo4j/graphalgo/impl/AllShortestPaths.java

```diff
@@ -0,0 +1,142 @@
+package org.neo4j.graphalgo.impl;
+
+import org.neo4j.graphalgo.api.Graph;
+import org.neo4j.graphalgo.core.utils.queue.IntMinPriorityQueue;
+import org.neo4j.graphdb.Direction;
+import org.neo4j.kernel.impl.util.collection.SimpleBitSet;
+
+import java.util.Arrays;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.LongStream;
+import java.util.stream.Stream;
+
+public class AllShortestPaths {
+
+    private final Graph graph;
+    private final int nodeCount;
+    private final int concurrency;
+    private final AtomicInteger counter;
+    private final ExecutorService executorService;
+    private final BlockingQueue<Result> resultQueue;
+
+    private boolean running;
+
+    public AllShortestPaths(Graph graph, ExecutorService executorService, int concurrency) {
+        this.graph = graph;
+        this.nodeCount = graph.nodeCount();
+        this.executorService = executorService;
+        if (concurrency < 1) {
+            throw new IllegalArgumentException("concurrency must be >0");
+        }
+        this.concurrency = concurrency;
+        this.counter = new AtomicInteger();
+        this.resultQueue = new LinkedBlockingDeque<>(); // TODO limit size?
+    }
+
+
+    public Stream<Result> resultStream() {
+
+        counter.set(0);
+        running = true;
+
+        for (int i = 0; i < concurrency; i++) {
+            executorService.submit(new ShortestPathTask());
+        }
+
+        long end = (long) nodeCount * nodeCount;
+
+        return LongStream.range(0, end)
```

Comment:
The stream is bounded at any rate, so close is more like an early terminate, not a required thing